### DISCUSSION TITLE
Two viscoelastic subdomains: material properties declared on block by multiple materials

### DISCUSSION THREAD

**Discussion by rynomcavoy** [idaholab/moose#28018] [Q&A Modules: Solid mechanics]
After looking at some related tickets: [22281](https://github.com/idaholab/moose/discussions/22281), etc., I still don't know how to proceed. I'm trying to run a file of uniaxial tension in a viscoelastic rectangle with a small region inside of it having different material parameters:

```
[GlobalParams]
  displacements = 'disp_x disp_y'
  #density = 1e4
[]

[Mesh]
  [generated]
    type = GeneratedMeshGenerator
    dim = 2
    nx = 40
    ny = 20
    xmax = 2
    ymax = 1
  []
  [block1]
    type = SubdomainBoundingBoxGenerator
    input = generated
    block_id = 1
    bottom_left = '0.9 0.3 0'
    top_right = '1.1 0.7 0'
    location = OUTSIDE
  []
  [block2]
    type = SubdomainBoundingBoxGenerator
    input = block1
    block_id = 2
    bottom_left = '0.9 0.3 0'
    top_right = '1.1 0.7 0'
    location = INSIDE
  []
[]

[Physics/SolidMechanics/QuasiStatic]
  [all]
    strain = SMALL
    add_variables = true
    generate_output = 'stress_xx stress_yx stress_yy vonmises_stress'
  []
[]

[BCs]
  [left_x]
    type = DirichletBC
    variable = disp_x
    boundary = left
    value = 0
  []
  [left_y]
    type = DirichletBC
    variable = disp_y
    boundary = left
    value = 0
  []
  [right_y]
    type = DirichletBC
    variable = disp_y
    boundary = right
    value = 0
  []
  [right_x]
    type = FunctionNeumannBC
    variable = disp_x
    boundary = right
    function = '600e6 * 0.2 * (1 + sin(2 * pi * t * 5))'
  []
[]

[Materials]
  [outer_mat]
    type = GeneralizedKelvinVoigtModel
    creep_modulus = 660e6
    creep_viscosity = 1
    young_modulus = 660e6
    poisson_ratio = 0.3
    block = 1
  []
  [inner_mat]
    type = GeneralizedKelvinVoigtModel
    creep_modulus = 100e4
    creep_viscosity = 2
    young_modulus = 100e4
    poisson_ratio = 0.3
    block = 2
  []
  [stress]
    type = ComputeMultipleInelasticStress
    inelastic_models = 'creep'
  []
  [creep]
    type = LinearViscoelasticStressUpdate
  []
  [strain]
    type = ComputeSmallStrain
    displacements = 'disp_x disp_y'
  []
[]

[UserObjects]
  [update_out]
    type = LinearViscoelasticityManager
    viscoelastic_model = 'outer_mat'
  []
  [update_inner]
    type = LinearViscoelasticityManager
    viscoelastic_model = 'inner_mat'
  []
[]

[Executioner]
  type = Transient
  solve_type = NEWTON
  petsc_options_iname = '-pc_type'
  petsc_options_value = 'lu'
  end_time = 1
  dt = 0.01
[]

[Outputs]
  exodus = true
[]
```

But an error occurs:
```
 `*** ERROR ***
The following error occurred in the Problem 'MOOSE Problem' of type FEProblem.

The following material properties are declared on block 1 by multiple materials:
  Material Property             Material Objects
  mechanical_strain             all_strain strain 
  total_strain                  all_strain strain `
```
I tried to use base_names e.g. [here](https://github.com/idaholab/moose/discussions/22714) but that only gave more errors. Any help appreciated.

**Comment by GiudGiud** [comment]
Hello

The 'strain' material you defined overlaps with the 'strain' material that is created by the Physics
```
[Physics/SolidMechanics/QuasiStatic]
  [all]
    strain = SMALL
    add_variables = true
    generate_output = 'stress_xx stress_yx stress_yy vonmises_stress'
  []
[]
```

I dont think you need both. I think you could delete the material and select the strain model in the Physics block

Guillaume

**Reply by rynomcavoy** [reply]
Commenting out the strain material gives a separate error:

````
*** ERROR ***
The following error occurred in the Problem 'MOOSE Problem' of type FEProblem.

Material property 'strain_increment', requested by 'stress' is not defined on block 1
Material property 'rotation_increment', requested by 'stress' is not defined on block 1
Material property 'strain_increment', requested by 'stress' is not defined on block 1
Material property 'strain_increment', requested by 'stress' is not defined on block 2
Material property 'rotation_increment', requested by 'stress' is not defined on block 2
Material property 'strain_increment', requested by 'stress' is not defined on block 2
```

**Reply by GiudGiud** [reply]
Can you make the physics use an incremental strain model?

Those properties are only defined for incremental models

**Reply by rynomcavoy** [reply]
No the Physics module only allows small or finite as inputs:

`*** ERROR ***
Invalid option "INCREMENTAL" in MooseEnum.  Valid options (not case-sensitive) are "SMALL FINITE".`

**Reply by GiudGiud** [reply]
set incremental = true

**Reply by rynomcavoy** [reply]
I went back to using the normal mechanics kernel and the simpler [LinearViscoelasticStress](https://mooseframework.inl.gov/source/materials/ComputeLinearViscoelasticStress.html) and a new error message is occuring:

`*** ERROR ***
Incompatible material and object blocks:
    /home/rmcavoy/projects/raptor/tendon/cgpt.i:102: (Materials/outer_mat/block) material defined on blocks 1, 
    block: object needs material on blocks 1, 2, `

Relevant input file:

`
[Mesh]
  [generated]
    type = GeneratedMeshGenerator
    dim = 2
    nx = 40
    ny = 20
    xmax = 2
    ymax = 1
  []
  [block1]
    type = SubdomainBoundingBoxGenerator
    input = generated
    block_id = 1
    bottom_left = '0.9 0.3 0'
    top_right = '1.1 0.7 0'
    location = OUTSIDE
  []
  [block2]
    type = SubdomainBoundingBoxGenerator
    input = block1
    block_id = 2
    bottom_left = '0.9 0.3 0'
    top_right = '1.1 0.7 0'
    location = INSIDE
  []
[]

[Variables]
  [./disp_x]
    order = FIRST
    family = LAGRANGE
  [../]
  [./disp_y]
    order = FIRST
    family = LAGRANGE
  [../]
[]

[Kernels]
  [SolidMechanics]
    displacements = 'disp_x disp_y'
    use_displaced_mesh = true
  [../]
[]

[BCs]
  [left_x]
    type = DirichletBC
    variable = disp_x
    boundary = left
    value = 0
  []
  [left_y]
    type = DirichletBC
    variable = disp_y
    boundary = left
    value = 0
  []
  [right_y]
    type = DirichletBC
    variable = disp_y
    boundary = right
    value = 0
  []
  [right_x]
    type = FunctionNeumannBC
    variable = disp_x
    boundary = right
    function = '600e6 * 0.2 * (1 + sin(2 * pi * t * 5))'
  []
[]

[Materials]
  [outer_mat]
    type = GeneralizedKelvinVoigtModel
    creep_modulus = 660e6
    creep_viscosity = 1
    young_modulus = 660e6
    poisson_ratio = 0.3
    block = 1
  []
  [inner_mat]
    type = GeneralizedKelvinVoigtModel
    creep_modulus = 100e4
    creep_viscosity = 2
    young_modulus = 100e4
    poisson_ratio = 0.3
    block = 2
  []
  [stress]
    type = ComputeLinearViscoelasticStress
  []
  [strain]
    type = ComputeSmallStrain
    displacements = 'disp_x disp_y'
  []
[]

[UserObjects]
  [update_out]
    type = LinearViscoelasticityManager
    viscoelastic_model = 'outer_mat'
  []
  [update_inner]
    type = LinearViscoelasticityManager
    viscoelastic_model = 'inner_mat'
  []
[]

[Executioner]
  type = Transient
  solve_type = NEWTON
  petsc_options_iname = '-pc_type'
  petsc_options_value = 'lu'
  end_time = 1
  dt = 0.01
[]

[Outputs]
  exodus = true
[]
`

**Reply by rynomcavoy** [reply]
> set incremental = true

Still, nope:

`*** ERROR ***
Incompatible material and object blocks:
    /home/rmcavoy/projects/raptor/tendon/cgpt.i:85: (Materials/outer_mat/block) material defined on blocks 1, 
    block: object needs material on blocks 1, 2, `

**Reply by GiudGiud** [reply]
the property defined by 'outer_mat' is being request on two blocks '1,2' when it only defines it on block 1
maybe you are missing another material to define the same property as outer_mat?

**Reply by GiudGiud** [reply]
> > set incremental = true
> 
> Still, nope:
> 
> `*** ERROR *** Incompatible material and object blocks: /home/rmcavoy/projects/raptor/tendon/cgpt.i:85: (Materials/outer_mat/block) material defined on blocks 1, block: object needs material on blocks 1, 2, `

Is this from your new input or from the input with the Physics?

**Reply by rynomcavoy** [reply]
Input with Physics
